Classes/EventStore/Storage/Doctrine/DoctrineEventStorage.php

```diff
@@ -117,18 +117,18 @@ public function load(StreamName $streamName, int $minimumSequenceNumber = 0): Ev
 
     /**
      * @inheritdoc
-     * @throws DBALException | ConcurrencyException
+     * @throws DBALException | ConcurrencyException | InvalidArgumentException | RuntimeException
      */
     public function commit(StreamName $streamName, WritableEvents $events, int $expectedVersion = ExpectedVersion::ANY): void
     {
         if ($streamName->isVirtualStream()) {
             throw new \InvalidArgumentException(sprintf('Can\'t commit to virtual stream "%s"', $streamName), 1540632984);
         }
         $this->reconnectDatabaseConnection();
-        $this->connection->beginTransaction();
-        if ($this->connection->getTransactionNestingLevel() > 1) {
+        if ($this->connection->getTransactionNestingLevel() > 0) {
```

Comment:
I would also like to challenge this check. What is the rationale to avoid committing if a transaction is already started? It is a feature available on a RDBMS which may have use cases (e.g. you want to make sure your command handler transactionally reads current state and commits to have stricter consistency guarantees) and I see no reason why this should be prevented at all.
As long as this code makes sure that for every `beginTransaction` a `commit` or a `rollback` happens, it is safe to be wrapped in another transaction. This is now the case.

Reply:
Good question. I'm not 100% sure about this one but wouldn't it be unexpected if your transaction was committed/rolled back from a "3rd party" if you publish an event half-way before?
I.e.:

1. start transaction
2. do changes
3. publish event(s)
4. (optionally) do more changes
5. commit transaction

should IMO be changed to

1. start transaction
2. do changes
3. (optionally) do more changes
4. commit transaction
5. publish event(s)

..or you might end up in an inconsistent (or at least unexpected) state.

Reply:
This is right, and of course the events table/store is the "source of truth", so nothing should go to the outside unless the source of truth is committed. The concern for publishing could be solved as above, or generally by the "outbox pattern". One of the reasons I don't like EventStores actively publishing changes ;) but I see it's relevant for some cases (minimize read model latency)

Reply:
Note that this is only relevant if your code uses the same instance of the doctrine connection as the event store adapter.